### ISSUE TITLE
Joining two tables retrieved from DB breaks Ballerina compilation

### ISSUE THREAD

**Issue by pasanbox** [ballerina-platform/ballerina-lang#10801]
**Description:**
I'm attempting to join two tables which I retrieved from a DB connection. When running this, Ballerina stops compilation and complains "**something went bad, sad**". Here is the contents of **ballerina-internal.log**
Can you please suggest what's wrong in my code? (Didn't ask on Stack Overflow because there's no clear compile error :( )
```
[2018-10-12 17:19:07,452] ERROR {org.ballerinalang.launcher.Main} - null 
java.util.NoSuchElementException
	at java.util.ArrayDeque.removeFirst(ArrayDeque.java:285)
	at org.wso2.ballerinalang.compiler.parser.BLangPackageBuilder.completeMatchNode(BLangPackageBuilder.java:2224)
	at org.wso2.ballerinalang.compiler.parser.BLangParserListener.exitMatchStatement(BLangParserListener.java:1264)
	at org.wso2.ballerinalang.compiler.parser.antlr4.BallerinaParser$MatchStatementContext.exitRule(BallerinaParser.java:6627)
	at org.antlr.v4.runtime.Parser.triggerExitRuleEvent(Parser.java:430)
	at org.antlr.v4.runtime.Parser.exitRule(Parser.java:646)
	at org.wso2.ballerinalang.compiler.parser.antlr4.BallerinaParser.compilationUnit(BallerinaParser.java:483)
	at org.wso2.ballerinalang.compiler.parser.Parser.generateCompilationUnit(Parser.java:109)
	at org.wso2.ballerinalang.compiler.parser.Parser.lambda$parse$0(Parser.java:84)
	at java.util.ArrayList.forEach(ArrayList.java:1257)
	at org.wso2.ballerinalang.compiler.parser.Parser.parse(Parser.java:84)
	at org.wso2.ballerinalang.compiler.PackageLoader.parse(PackageLoader.java:421)
	at org.wso2.ballerinalang.compiler.PackageLoader.loadEntryPackage(PackageLoader.java:291)
	at org.wso2.ballerinalang.compiler.Compiler.lambda$compilePackages$3(Compiler.java:147)
	at java.util.stream.ReferencePipeline$3$1.accept(ReferencePipeline.java:193)
	at java.util.stream.ReferencePipeline$2$1.accept(ReferencePipeline.java:175)
	at java.util.stream.Streams$StreamBuilderImpl.forEachRemaining(Streams.java:419)
	at java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:481)
	at java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:471)
	at java.util.stream.ReduceOps$ReduceOp.evaluateSequential(ReduceOps.java:708)
	at java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:234)
	at java.util.stream.ReferencePipeline.collect(ReferencePipeline.java:499)
	at org.wso2.ballerinalang.compiler.Compiler.compilePackages(Compiler.java:149)
	at org.wso2.ballerinalang.compiler.Compiler.compilePackage(Compiler.java:173)
	at org.wso2.ballerinalang.compiler.Compiler.compile(Compiler.java:85)
	at org.wso2.ballerinalang.compiler.Compiler.compile(Compiler.java:76)
	at org.ballerinalang.launcher.LauncherUtils.compile(LauncherUtils.java:299)
	at org.ballerinalang.launcher.LauncherUtils.runProgram(LauncherUtils.java:118)
	at org.ballerinalang.launcher.Main$RunCmd.execute(Main.java:277)
	at java.util.Optional.ifPresent(Optional.java:159)
	at org.ballerinalang.launcher.Main.main(Main.java:65)
```
 

**Steps to reproduce:**
Please find the code [here](https://gist.github.com/pasanbox/faf1207c7b83e62e539e7fda632affb8).
SQL to create test tables: [table_join.sql.zip](https://github.com/ballerina-platform/ballerina-lang/files/2472914/table_join.sql.zip)

**Affected Versions:**
Ballerina 0.982.0

**OS, DB, other environment details and versions:**
macOS 10.13.6

**Related Issues (optional):**

**Suggested Labels (optional):**

**Suggested Assignees (optional):**

**Comment by Manuri** [comment]
@pasanbox  ATM this is only supported for table literals, but not for the tables obtained through a db client endpoint. Can you perform the join at the database level for this?

**Comment by pasanbox** [comment]
Hi @Manuri, Thanks for the clarification.
The tables I actually have to join belong to two databases. Is there a way to do the joining at DB level in that case ?

**Comment by Manuri** [comment]
According to the given sample you are using mysql. If they are in the same server, I think that should be possible with adding the database name qualifier to the table name. 
eg: db1.table1
Have you tried that?

**Comment by pasanbox** [comment]
Hi @Manuri, I've tried that and it works. Thanks.
So is it safe to assume the 'name' config in ClientEndpointConfiguration is optional and we can connect to many databases in a single host using a single mysql client endpoint?
I'm asking because the [documentation](https://ballerina.io/learn/api-docs/ballerina/mysql.html) suggests the endpoint is targetted for a single database, and not multiple databases... 

> virtual representation of the physical endpoint of the MySQL database that you are trying to connect to

**Comment by Manuri** [comment]
If the database supports this, then its fine to map an endpoint with a server rather than a single database. In this case since mysql supports it, it is safe to do so.

**Comment by pasanbox** [comment]
@Manuri Thanks a lot for the explanation.

**Comment by Manuri** [comment]
Keeping this issue open as it is required to fix bad-sad error and provide a proper error message to the user.

**Comment by Manuri** [comment]
@uthaiyashankar [This ](https://ballerina.io/learn/by-example/table-queries.html) is where we demonstrate the capabilities of table queries. Actually we haven't added to documentation that it is not possible to do this for tables obtained through database clients. We'll add that.

Actually this is supported only for in-memory tables, as decided from the beginning. There was an initial idea to make it possible to join tables obtained even through different database client endpoints, but as of now that is not included in the immediate plans.

**Comment by nimanthag** [comment]
@Manuri Could you please explain the connection pool objects behavior in the above scenario ? For my knowledge maximum pool size configuration is available per endpoint (not per server).

**Comment by Manuri** [comment]
@nimanthag For each endpoint we internally create a connection pool. The size of this pool is configurable. Whether we can use this endpoint to query only a given database or any database in the same server depends on whether that database server supports it or not.

**Comment by NatashaWso2** [comment]
@pasanbox closing this issue since it has been already fixed with https://github.com/ballerina-platform/ballerina-lang/pull/11944. The fix will is available with [ Ballerina-0.990.0](https://ballerina.io/downloads/)
 Please reopen if you think there are any other areas to be fixed or improved.
